Add single_branch option in git module.

zehome opened the pull request:
##### SUMMARY

In some usecases, we want to be able to clone a single branch
of a repository, without using --depth (which implies --single-branch).

Fixes https://github.com/ansible/ansible-modules-core/issues/1463

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
source_control/git

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (git_single_branch 127b28cbab) last updated 2017/08/21 11:16:53 (GMT +200)
  config file = None
  configured module search path = [u'/home/ed/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/ed/ansible/lib/ansible
  executable location = /home/ed/ansible/bin/ansible
  python version = 2.7.13+ (default, Jul 19 2017, 18:15:03) [GCC 6.4.0 20170704]
```

Commit:
Add single_branch option in git module.

In some usecases, we want to be able to clone a single branch
of a repository, without using --depth (which implies --single-branch).

robinro commented:
@zehome Thanks for this PR.

Can you add an integration test for this flag?

What is the behavior when there are updates? The module should be idempotent, so running it a second time after a clone should not fetch anything new. To do this correctly one can probably reuse some of the code from `--depth`:
https://github.com/zehome/ansible/blob/127b28cbab3088d7af467fa0798d96e148ca528a/lib/ansible/modules/source_control/git.py#L719-L743

zehome commented:
Ok, I'll work on it a little bit more.

nitzmahone commented:
This seems like it'd be a useful feature for overly large repos, but definitely needs tests. Anyone interested in writing some?

samdoran committed:
Use branch name when available

- update description of parameter
- consolidate branch or tag checking for easy reuse

samdoran committed:
Add changelog

samdoran committed:
Use static task imports rather than dynamic includes

samdoran committed:
Add integration tests for single_branch

samdoran commented:
Need to handle older versions of `git` that don't have this option.

samdoran committed:
Account for older versions of git

samdoran committed:
Minor tweak to warnings
